var/spack/repos/builtin/packages/gcc/package.py

```diff
@@ -82,25 +94,75 @@ class Gcc(AutotoolsPackage):
     # depends_on('autogen@5.5.4:', type='test')
     # depends_on('guile@1.4.1:', type='test')
 
-    if sys.platform == 'darwin':
-        patch('darwin/gcc-4.9.patch1', when='@4.9.0:4.9.3')
-        patch('darwin/gcc-4.9.patch2', when='@4.9.0:4.9.3')
-    else:
-        provides('golang', when='@4.7.1:')
-
+    # See https://golang.org/doc/install/gccgo#Releases
+    provides('golang', when='languages=go')
+    # 'when' does not currently support multiple parts of a spec.
+    # See https://github.com/LLNL/spack/pull/4151
+    # provides('golang',        when='languages=go @4.6:')
+    # provides('golang@:1',     when='languages=go @4.7.1:')
+    # provides('golang@:1.1',   when='languages=go @4.8:')
+    # provides('golang@:1.1.2', when='languages=go @4.8.2:')
+    # provides('golang@:1.2',   when='languages=go @4.9:')
+    # provides('golang@:1.4',   when='languages=go @5:')
+    # provides('golang@:1.6.1', when='languages=go @6:')
+    # provides('golang@:1.8',   when='languages=go @7:')
+
+    patch('darwin/gcc-7.1.0-headerpad.patch', when='@7.1.0 platform=darwin')
+    patch('darwin/gcc-6.1.0-jit.patch', when='@6.1.0 platform=darwin')
+    patch('darwin/gcc-4.9.patch1', when='@4.9.0:4.9.3 platform=darwin')
+    patch('darwin/gcc-4.9.patch2', when='@4.9.0:4.9.3 platform=darwin')
     patch('piclibs.patch', when='+piclibs')
     patch('gcc-backport.patch', when='@4.7:4.9.2,5:5.3')
 
+    @run_before('autoreconf')
+    def check_languages(self):
+        """Makes sure all requested languages are valid for a
+        specific version of GCC.
+
+        For a list of valid languages for a specific release,
+        run the following command in the GCC source directory:
+
+        .. code-block:: console
+
+           $ grep ^language= gcc/*/config-lang.in
+
+        See https://gcc.gnu.org/install/configure.html
+        """
+        spec = self.spec
+        version = self.version
+
+        # Support for processing BRIG 1.0 files was added in GCC 7
+        # BRIG is a binary format for HSAIL:
+        # (Heterogeneous System Architecture Intermediate Language).
+        # See https://gcc.gnu.org/gcc-7/changes.html
+        if version < Version('7') and 'languages=brig' in spec:
```

Comment:
maybe `conflict()` statement should indeed be extended to provide custom message?
This could read
```
conflicts('languages=brig', when='@:6.9999','BRIG is not available before GCC 7')
```

My understand that `conflict()` will fail during conretization, whereas the checks here will fail during installation. 
Generally i think we should stick to the former to avoid waisting time and installing half of the DAG to fail at some node.